doc: Cheat Sheet (issue #4316) (#5458)
microsoft/fast: sites/website/src/docs/resources/cheat-sheet.md

---
id: cheat-sheet
title: Cheat Sheet
sidebar_label: Cheat Sheet
custom_edit_url: https://github.com/microsoft/fast/edit/master/sites/website/src/docs/resources/cheat-sheet.md
---

# Cheat Sheet

## Packages:

### [@microsoft/fast-element](https://www.fast.design/docs/fast-element/getting-started)

> **A library that is a lightweight means to easily build performant, memory-efficient, standards-compliant Web Components.**

Reviewer: Can we reword this? It reads a bit off, though I think I understand the intent. Specifically, I think "A library that is. lightweight means to..." could probably be reworked to be more easily consumed.

Author: @EisenbergEffect When you have a moment, I am not sure how to best reword this sentence.

Should I include some of the content from the intro page from the docs?

![Screen Shot 2021-12-15 at 2 18 28 PM](https://user-images.githubusercontent.com/35910788/146273747-dfd57ab8-758e-4f72-9632-58cd90fef783.png)

This sentence is found in the docs and in the repo readme.  Would it make sense to update in these two places as well?

![Screen Shot 2021-12-15 at 2 10 40 PM](https://user-images.githubusercontent.com/35910788/146272899-47f61ad8-9879-4c4c-a398-a54cc80ad4fb.png)

![Screen Shot 2021-12-15 at 2 10 30 PM](https://user-images.githubusercontent.com/35910788/146272934-7069fc94-a17f-4d96-9e1b-151b85d03356.png)

Reviewer 2: Perhaps something like "A lightweight library for building performant, memory-efficient, standards-compliant Web Components."

Author: @EisenbergEffect Thank you - updated in the cheat sheet!

> * Provides a thin layer of opinions on top of [Web Components](https://www.fast.design/docs/resources/why-web-components), lifting the level of abstraction just enough to make it easier and FASTer to [build components]((#building-components)).   

Reviewer: ```suggestion
* Provides a thin layer of opinion on top of [Web Components](https://www.fast.design/docs/resources/why-web-components), lifting the level of abstraction just enough to make it easier and faster to [build components]((#building-components)).
```

Just a suggestion - I like the nod to "FAST" with 'faster', but I'm not sure we want to make that a convention. Open to others feedback on that though.

Author: @chrisdholt That makes sense, thank you.  I grabbed the verbiage from the docs: https://www.fast.design/docs/introduction.  I can make the updates here as well:

<img width="1004" alt="Screen Shot 2021-12-14 at 11 36 44 AM" src="https://user-images.githubusercontent.com/35910788/146071196-71a0340b-eda0-4983-aa5e-5a20ecb4df8d.png">

* Use this library when you want to create new custom web components. 

To install the fast-element library, use either `npm` or `yarn`:

```shell
npm install --save @microsoft/fast-element
```

```shell
yarn add @microsoft/fast-element
```

Within your JavaScript or TypeScript code, you can then import library APIs like this:

```ts
import { FASTElement } from "@microsoft/fast-element";
```

---

### [@microsoft/fast-foundation](https://www.fast.design/docs/introduction)

> **A library that provides foundational building blocks that can be assembled to create new design systems and component libraries.**

Reviewer: Technically I think this still leaves a knowledge gap that I'd love to figure out how exactly to close. This is the "larger" and likely most common purpose, but Foundation also provides a set of common primitive classes and templates which represent functional elements. In the case of something like a "color swatch", Foundation gives us reusable radio/radiogroup bits which can then be styled to meet this scenario. I don't think there's anything wrong with just focusing on this portion, but Foundation enables so much and it seems like the quick hit here is scoped to two things (which area kind of one most of the time...).

Author: @chrisdholt Thank you - I updated with verbiage from here:

![Screen Shot 2021-12-15 at 4 32 13 PM](https://user-images.githubusercontent.com/35910788/146286242-7adcc5f5-5fd3-48ea-b9a8-dbbcc50da288.png)

* Exports of this package can generally be thought of as un-styled base components that implement semantic and accessible markup and behavior.   

> * Use this library when you want to implement something like Google's Material Design or Twitter Bootstrap.

Reviewer: Is implement the right word here - perhaps "create" better convey's the intent? One other question on this is if Twitter is still associated with the latest versions of the project; I'm not sure they are. I did a quick search specifically for "Twitter Bootstrap" and it linked to _pinned_ versions of the project. The latest versions (4/5) don't seem to call out that relationship at all. I wonder if we can leave off that affiliation:

```suggestion
* Use this library when you want to create something like Google's Material Design or Bootstrap.
```

Author: @chrisdholt Thank you - updated!  I also updated the verbiage from the intro page: https://www.fast.design/docs/introduction

<img width="975" alt="Screen Shot 2021-12-14 at 2 26 04 PM" src="https://user-images.githubusercontent.com/35910788/146089266-8b031a4f-61d6-4f6b-ab9d-e49b7acf5c46.png">

```ts
import {
  Button as FoundationButton,
  buttonTemplate as template,
} from "@microsoft/fast-foundation";
import { css } from "@microsoft/fast-element";

export class MyButton extends FoundationButton {
  ...
}

export const buttonStyles = css`
  :host {
    background-color: azure;
  }
`;

export const myButton = MyButton.compose({
  baseName: "button",
  baseClass: FoundationButton,
  template,
  styles: buttonStyles,
  shadowOptions: {
    delegatesFocus: true,
  },
});
```

---

### [@microsoft/fast-components](https://www.fast.design/docs/components/getting-started)     

> **A component library that implements Microsoft's [FAST Frame Design System](https://www.fast.design/docs/design-systems/fast-frame/).**

Reviewer: "FAST Frame" has been what we've referred to the system as - while the project is under the Microsoft umbrella, the system _is_ FAST components. I'd like to rework this if we can. Perhaps:

**The FAST component library and visual system [FAST Frame](https://www.fast.design/docs/design-systems/fast-frame/), which implements fast-foundation.**

Author: @chrisdholt Thank you - updated!

* Assembles the building blocks of `@microsoft/fast-foundation` to create its component set. 
* Use this library when you want to integrate [FAST Components](#using-components) into an existing site or app.

To install the components, use either `npm` or `yarn`:

```shell
npm install --save @microsoft/fast-components
``` 

```shell
yarn add @microsoft/fast-components
```

> To use a Web Component as a custom element in HTML, the custom element must be registered:

Reviewer: This seems generic but we're talking specifically about the library in this case. Perhaps instead of "To use a Web Component" here we can be more specific to the package?

```suggestion
To use a component as a custom element in HTML, the custom element must be registered.
```

Author: @chrisdholt Thank you - updated!

```ts
import { 
  fastButton, 
  provideFASTDesignSystem 
} from "@microsoft/fast-components";

provideFASTDesignSystem()
  .register(

>     fastButton() // custom element

Reviewer: @eisenbergeffect technically this isn't the custom element itself - is it confusing if we conflate the two? If our intent is a quick look and education I could see where this doesn't hurt, but I could also see where it causes confusion in terminology.

Reviewer 2: I think we should adjust the wording regardless. It could just be changed to "custom element registration"

Author: @chrisdholt @EisenbergEffect Thank you - updated!  

I am also not opposed to removing the comment.

* Assembles the building blocks of `@microsoft/fast-foundation` to create its component set. 
* Use this library when you want components to look and feel like those found in Microsoft products.

To install the components, use either `npm` or `yarn`:

```shell
npm install --save @fluentui/web-components
``` 

```shell
yarn add @fluentui/web-components
```

To use a Web Component as a custom element in HTML, the custom element must be registered:

```ts
import { 
  fluentButton, 
  provideFluentDesignSystem 
} from "@fluentui/web-components";

provideFluentDesignSystem()
  .register(
    fluentButton() // custom element
  );
```

---

If you are attempting to configure the components for integration into a specific Microsoft product, the following table provides `AccentBaseColor` values you can use:

| Product     | AccentBaseColor | 
| :---        | :---            | 
| Office      | #D83B01         | 
| Word        | #185ABD         | 
| Excel       | #107C41         | 
| PowerPoint  | #C43E1C         | 
| Teams       | #6264A7         | 
| OneNote     | #7719AA         | 
| SharePoint  | #03787C         | 
| Stream      | #BC1948         | 

> ---

Reviewer: @EisenbergEffect @huynhicode I think we should remove this - while we're the primary maintainers of Fluent UI Web Components as well, there is an entire documentation strategy there and the code _doesn't_ live here. I think adding this here is going to cause more work for us in managing where issues should be raised and where support should come from. 

My suggestion is that if we want to add this, we should add it to Fluent. Alterntively, we could have a "projects" portion where we call out _all_ partners and provide more context. As it stands right now, this is an implementation we just happen to be more involved with than not, but it's not part of FAST - it's an implementation of it. It's especially odd as we jump back into FAST below.

Reviewer 2: So, maybe remove the chart and the accent color note above that an dreplace it with a link to the Fluent UI Web Components documentation?

@Chris Do you have the repo link where I could contribute this over there? How are we managing that now?

Reviewer: Fork and contribute to our WC Readme is one option, we can also add to the msdocs site: https://github.com/microsoft/fluentui

Reviewer 2: I was thinking to add it to docs.microsoft.com docs, in the styling section.

Reviewer: I think that's probably a better place for the last part. The rest of this should be represented in the docs there.

Reviewer: @huynhicode just for clarity, this is really the primary thing I'm blocking on. Once this is removed I'm happy to approve given the other changes being made :)

Thanks for this!

Author: @chrisdholt Make sense, thank you!  I appreciate all the feedback!

Author: @chrisdholt @EisenbergEffect Thank you - updated!  I removed the table and corresponding table note, and added a link to Fluent UI Web.

FAST libraries can also be used in combination with a wide variety of existing technologies.

- [Angular](https://www.fast.design/docs/integrations/angular)
- [ASP.NET](https://www.fast.design/docs/integrations/aspnet)
- [Aurelia](https://www.fast.design/docs/integrations/aurelia)
- [Blazor](https://www.fast.design/docs/integrations/blazor)
- [Ember](https://www.fast.design/docs/integrations/ember)
- [React](https://www.fast.design/docs/integrations/react)
- [Vue](https://www.fast.design/docs/integrations/vue)
- [Webpack](https://www.fast.design/docs/integrations/webpack)

Not seeing an integration for your preferred technology?  Open an issue on [GitHub](https://github.com/microsoft/fast/issues).

---

## [Building Components](https://www.fast.design/docs/fast-element/getting-started)

### Setup

To define a custom element:

```ts
import { FASTElement, customElement } from "@microsoft/fast-element";

@customElement("name-tag") // custom element being created
export class NameTag extends FASTElement {

}
```

With this in place, you can now use your `<name-tag>` element anywhere in HTML with the following markup:

```html
<name-tag></name-tag>
```

> 

Reviewer: @eisenbergeffect @huynhicode this isn't exhaustive as we have this AND the registration portion, which depending on use case/scenario, either seems reasonable. If I were doing a quick run through this, I would _assume_ this is "the way". How do we capture both in the context of this cheat sheet? Would perhaps linking to something be more effective?

Reviewer 2: I think we could have a short explanation here that says there are two main approaches. The first approach is for simple declarations of non-shared components. The second style is for components designed to be published in shareable libraries. We could link down in the doc to the second approach if it makes sense to put that one somewhere else.

Author: @chrisdholt @EisenbergEffect Thank you - I updated the "Building Components" section with the additional verbiage and a link to the Design Systems section.

```shell
yarn add @microsoft/fast-components
```

To use a Web Component as a custom element in HTML, the custom element must be registered:

```ts
import { 
    fastButton, 
    provideFASTDesignSystem 
} from "@microsoft/fast-components";

provideFASTDesignSystem()
    .register(
        fastButton() // custom element
    );
```

---
### [@fluentui/web-components](https://www.fast.design/docs/components/getting-started)

**A component library that implements Microsoft's [Fluent Design System](https://www.microsoft.com/design/fluent/#/).**

* Assembles the building blocks of `@microsoft/fast-foundation` to create its component set. 

* Use this library when you want components to look and feel like those found in Microsoft products.

To install the components, use either `npm` or `yarn`:
```shell
npm install --save @fluentui/web-components
``` 
```shell
yarn add @fluentui/web-components
```

To use a Web Component as a custom element in HTML, the custom element must be registered:

```ts
import { 
    fluentButton, 
    provideFluentDesignSystem 
} from "@fluentui/web-components";

provideFluentDesignSystem()
    .register(
        fluentButton() // custom element
    );
```

---

> If you are attempting to configure the components for integration into a specific Microsoft product, the following table provides `AccentBaseColor` values you can use:

Reviewer 3: This should link to the "Configuring a design token" section and/or provide an example like:
`accentPalette.withDefault(PaletteRGB.create(SwatchRGB.from(parseColorHexRGB(hexColor)!)));`

"AccentBaseColor" isn't in code yet. There is a PR adding support for that, but there are some open design questions we need to resolve.

Author: @bheston Thank you - Fluent content has been removed from the cheat sheet per @chrisdholt 's request.

| Product     | AccentBaseColor | 
| :---        | :---            | 
| Office      | #D83B01         | 
| Word        | #185ABD         | 
| Excel       | #107C41         | 
| PowerPoint  | #C43E1C         | 
| Teams       | #6264A7         | 
| OneNote     | #7719AA         | 
| SharePoint  | #03787C         | 
| Stream      | #BC1948         | 

---

## [Using Components](https://www.fast.design/docs/components/getting-started)
### Setup

To register design system components:

```ts
import { 
    fastButton, 
    provideFASTDesignSystem 
} from "@microsoft/fast-components";

provideFASTDesignSystem()
    .register(
        fastButton() // desired custom component
    );
```

Or, register all system components:

```ts
import { 
    allComponents, 
    provideFASTDesignSystem 
} from "@microsoft/fast-components";

provideFASTDesignSystem().register(allComponents);
```

With the components registered, add any component to the HTML:

```html
<fast-button>Hello world</fast-button>
```

---

### [FAST Components](https://explore.fast.design/)

> | Component | Name | Component Explorer | 

Reviewer 3: This feels hard to maintain. I would just defer to an authoritative source of all components like Explorer or Storybook rather than list them all again (for now) here.

Reviewer 4: I agree, we list them several times in a few locations, they should only be located in one.

Author: @bheston @janechu Great points, thank you.  I will remove the table.

Would this make more sense?

Launch our [Component Explorer](https://explore.fast.design/) for a comprehensive list of FAST Components.

  ...
}

export const buttonStyles = css`
  :host {
    background-color: azure;
  }
`;

export const myButton = MyButton.compose({
  baseName: "button",
  baseClass: FoundationButton,
  template,
  styles: buttonStyles,
  shadowOptions: {
    delegatesFocus: true,
  },
});
```
---

### [@microsoft/fast-components](https://www.fast.design/docs/components/getting-started)     

**A component library that implements Microsoft's [FAST Frame Design System](https://www.fast.design/docs/design-systems/fast-frame/).**

* Assembles the building blocks of `@microsoft/fast-foundation` to create its component set. 

* Use this library when you want to integrate [FAST Components](#using-components) into an existing site or app.

To install the components, use either `npm` or `yarn`:
```shell
npm install --save @microsoft/fast-components
``` 
```shell
yarn add @microsoft/fast-components
```

To use a Web Component as a custom element in HTML, the custom element must be registered:

```ts
import { 
    fastButton, 
    provideFASTDesignSystem 
} from "@microsoft/fast-components";

provideFASTDesignSystem()
    .register(
        fastButton() // custom element
    );
```

---

> ### [@fluentui/web-components](https://www.fast.design/docs/components/getting-started)

Reviewer 4: Should we link to some documentation specifically for the fluent UI web components?

Author: @janechu Thank you - updated!

**A component library that implements Microsoft's [Fluent Design System](https://www.microsoft.com/design/fluent/#/).**

* Assembles the building blocks of `@microsoft/fast-foundation` to create its component set. 

> 

Reviewer 4: ```suggestion
```

Author: @janechu Thank you - removed!

| [fast-tooltip](https://www.fast.design/docs/components/tooltip) | `fastTooltip()` | [Tooltip](https://explore.fast.design/components/fast-tooltip) | 
| [fast-tree-view](https://www.fast.design/docs/components/tree-view) | `fastTreeItem()` `fastTreeView()` | [Tree view](https://explore.fast.design/components/fast-tree-view) | 

---
### [Integrations](https://www.fast.design/docs/integrations/introduction)

FAST libraries can also be used in combination with a wide variety of existing technologies.

- [Angular](https://www.fast.design/docs/integrations/angular)
- [ASP.NET](https://www.fast.design/docs/integrations/aspnet)
- [Aurelia](https://www.fast.design/docs/integrations/aurelia)
- [Blazor](https://www.fast.design/docs/integrations/blazor)
- [Ember](https://www.fast.design/docs/integrations/ember)
- [React](https://www.fast.design/docs/integrations/react)
- [Vue](https://www.fast.design/docs/integrations/vue)
- [Webpack](https://www.fast.design/docs/integrations/webpack)

Not seeing an integration for your preferred technology?  Open an issue on [GitHub](https://github.com/microsoft/fast/issues).

---

> ## [Building Components](https://www.fast.design/docs/components/getting-started)

Reviewer 4: ```suggestion
## [Building Components](https://www.fast.design/docs/fast-element/getting-started)
```

Author: @janechu Thank you - updated!

The `createBehavior()` method can be used to create a `Behavior` that is bound to the element using the `CSSDirective`:

```ts
class RandomWidth extends CSSDirective {
  private property = "--random-width";
  createCSS() {
    return `width: var(${this.property});`
  }

  createBehavior() {
    return {
      bind(el) {
        el.style.setProperty(this.property, Math.random() * 100)
      }
      unbind(el) {
        el.style.removeProperty(this.property);
      }
    }
  }
}
```

---
## Design System
### [Creating a Design System](https://www.fast.design/docs/design-systems/creating-a-component-library#defining-a-design-system)

1. Start with the base components defined in `@microsoft/fast-foundation` and compose them with your own styles:

```ts
import {
    Button,
    buttonTemplate as template,
} from "@microsoft/fast-foundation";
import { buttonStyles as styles } from "./special-button.styles";

export const specialButton = Button.compose({
    baseName: "button",
    template,
    styles,
    shadowOptions: {
        delegatesFocus: true,
    },
});

export const buttonStyles = styles;
```

> 
> 
> 

Reviewer 4: ```suggestion
```

Author: @janechu Thank you - removed!

---
id: cheat-sheet
title: Cheat Sheet
sidebar_label: Cheat Sheet
custom_edit_url: https://github.com/microsoft/fast/edit/master/sites/website/src/docs/resources/cheat-sheet.md
---

# Cheat Sheet

## Packages:

### [@microsoft/fast-element](https://www.fast.design/docs/fast-element/getting-started)

**A library that is a lightweight means to easily build performant, memory-efficient, standards-compliant Web Components.**

* Provides a thin layer of opinions on top of [Web Components](https://www.fast.design/docs/resources/why-web-components), lifting the level of abstraction just enough to make it easier and FASTer to [build components]((#building-components)).   

> * Use this library when you want to create completely new web components. 

Author: @janechu Thank you - updated!

To install the fast-element library, use either `npm` or `yarn`:

```shell
npm install --save @microsoft/fast-element
```

```shell
yarn add @microsoft/fast-element
```

Within your JavaScript or TypeScript code, you can then import library APIs like this:

```ts

> import { FASTElement } from '@microsoft/fast-element';

Author: @radium-v Thank you - updated!

* Assembles the building blocks of `@microsoft/fast-foundation` to create its component set. 

* Use this library when you want to integrate [FAST Components](#using-components) into an existing site or app.

To install the components, use either `npm` or `yarn`:
```shell
npm install --save @microsoft/fast-components
``` 
```shell
yarn add @microsoft/fast-components
```

To use a Web Component as a custom element in HTML, the custom element must be registered:

```ts
import { 
    fastButton, 
    provideFASTDesignSystem 
} from "@microsoft/fast-components";

provideFASTDesignSystem()
    .register(
        fastButton() // custom element
    );
```

---
### [@fluentui/web-components](https://www.fast.design/docs/components/getting-started)

**A component library that implements Microsoft's [Fluent Design System](https://www.microsoft.com/design/fluent/#/).**

* Assembles the building blocks of `@microsoft/fast-foundation` to create its component set. 

* Use this library when you want components to look and feel like those found in Microsoft products.

To install the components, use either `npm` or `yarn`:
```shell

> npm install --save @microsoft/fast-components

Reviewer 5: ```suggestion
npm install --save @fluentui/web-components
```

Author: Thank you @nicholasrice - updated!

``` 
```shell

> yarn add @microsoft/fast-components

Reviewer 5: ```suggestion
yarn add @fluentui/web-components
```

Author: Thank you @nicholasrice - updated!

| [fast-toolbar](https://www.fast.design/docs/components/toolbar)                           | `fastToolbar()`                                               | [Toolbar](https://explore.fast.design/components/fast-toolbar)                                                                                        |                                                                                                                                          
| [fast-tooltip](https://www.fast.design/docs/components/tooltip)                           | `fastTooltip()`                                               | [Tooltip](https://explore.fast.design/components/fast-tooltip)                                                                                        |                                                                                                                                             

> | [fast-tree-view](https://www.fast.design/docs/components/tree-view)                       | `fastTreeItem()` `fastTreeView()`                             | [Tree view](https://explore.fast.design/components/fast-tree-view)                                                                                    |                                                                                                                                      

Author: @radium-v Thank you - updated.

---
### [Integrations](https://www.fast.design/docs/integrations/introduction)

FAST libraries can also be used in combination with a wide variety of existing technologies.

[Angular](https://www.fast.design/docs/integrations/angular)

[ASP.NET](https://www.fast.design/docs/integrations/aspnet)

[Aurelia](https://www.fast.design/docs/integrations/aurelia)

[Blazor](https://www.fast.design/docs/integrations/blazor)

[Ember](https://www.fast.design/docs/integrations/ember)

[React](https://www.fast.design/docs/integrations/react)

[Vue](https://www.fast.design/docs/integrations/vue)

> [Webpack](https://www.fast.design/docs/integrations/webpack)

Author: @radium-v Thank you - updated!

Not seeing an integration for your preferred technology?  Open an issue on [GitHub](https://github.com/microsoft/fast/issues).

---
## [Building Components](https://www.fast.design/docs/components/getting-started)
### Setup

To define a custom element:

```ts
import { FASTElement, customElement } from '@microsoft/fast-element';

@customElement('name-tag') // custom element being created
export class NameTag extends FASTElement {

}
```

With this in place, you can now use your name-tag element anywhere in HTML with the following markup:

```html
<name-tag></name-tag>
```

---

### [Attributes](https://www.fast.design/docs/fast-element/defining-elements#customizing-attributes)

To add attributes to your HTML element, create properties decorated by the `@attr` decorator:

```ts
import { FASTElement, customElement, attr } from '@microsoft/fast-element';

@customElement('name-tag')
export class NameTag extends FASTElement {
  @attr greeting: string = 'Hello';

  greetingChanged() {

>     this.shadowRoot!.innerHTML = this.greeting;

Reviewer 6: Should our examples include setting "innerHTML" given that can have security issues?

Reviewer 6: Might also confuse someone into thinking this is how to render something into a component rather than a template binding.  Maybe the example could emit an event?

Reviewer: Relates to my comment above. There's a number of ways to leverage and use - how do we determine what should be highlighted as "the way" in the cheat sheet?

Author: @scomea @chrisdholt Thank you - updated!

  }
}
```

To use a Web Component with Attributes:

```html
<name-tag greeting="Hola"></name-tag>
```

---

#### [Customizing attributes](https://www.fast.design/docs/fast-element/defining-elements#customizing-attributes):

The three modes available through the `mode` property of the attribute configuration:

| Mode | Guidance |
| :-- | :-- | 
| reflect | The default mode that is used if none is specified. 
| boolean |  This mode causes your attribute to function using the HTML boolean attribute behavior.|
| fromView |  This mode skips reflecting the value of the property back to the HTML attribute.

---

> Supply a custom `ValueConverter` by setting the converter property of the attribute configuration:

Reviewer: It may help to use "ValueConverter" prior to this to link the concept to the above modes.

> Supply a custom `ValueConverter` by setting the converter property of the attribute configuration:

Author: @radium-v @chrisdholt Thank you - I added additional content to this section.

```ts
interface ValueConverter {
    toView(value: any): string;
    fromView(value: string): any;
}
```

Example:

```ts
const numberConverter: ValueConverter = {
  toView(value: number): string {
    return String(value);
  },

  fromView(value: string): number {
    return Number(value);
  }
};

@customElement('my-counter')
export class MyCounter extends FASTElement {

>   @attr({ mode: 'reflect', converter: numberConverter }) count: number = 0;

Author: @radium-v Thank you - updated!

}
```

---
### [Templates](https://www.fast.design/docs/fast-element/declaring-templates)

> To create an HTML template for an element, import and use the html tagged template helper and pass the template to the `@customElement` decorator:

Author: @radium-v Thank you - updated!

const template = html<NameTag>`
  <div class="header">
    <h3>${x => x.greeting.toUpperCase()}</h3>
    <h4>my name is</h4>
  </div>

  <div class="body">TODO: Name Here</div>

  <div class="footer"></div>
`;

@customElement({
  name: 'name-tag',
  template
})
export class NameTag extends FASTElement {
  ...
}
```

---

### [Observables](https://www.fast.design/docs/fast-element/observables-and-state#observable-features)

To enable binding tracking and change notification, properties must be decorated with either `@attr` or `@observable`. 

Use `@attr` for primitive properties (string, bool, number) that are intended to be surfaced on your element as HTML [attributes](#attributes). Use `@observable` for all other property types on an HTMLElement and all observable properties on plain classes.

```ts
import { Observable } from '@microsoft/fast-element';

export class Person {
  @observable firstName = '';
  @observable lastName = '';

  get fullName() {
    return `${this.firstName} ${this.LastName}`;
  }
}
```

---
### [Bindings](https://www.fast.design/docs/fast-element/declaring-templates#understanding-bindings)

| Binding Type | Example | Notes |
| :--- | :--- | :--- |

> | Content | `<a>${…}</a>` | Creates a binding to interpolate text or child templates into element content. |

Reviewer 7: The expressions in the examples should include arrow functions to emphasize the binding functionality

Author: @radium-v Thank you - updated!

Would you mind reviewing the updates in the Binding table examples?

Author: @EisenbergEffect would you mind reviewing the updates I made to the Binding table as well?  I updated the example code and added to the "event handler" notes section to include preventDefault() verbiage.

Reviewer 2: Looks good to me!

Author: @EisenbergEffect Thank you for reviewing!

| HTML Attribute | `<a href=${…}></a>` | Creates a binding that uses the setAttribute API. Attribute bindings also support interpolation with text and other bindings. |
| HTML Boolean Attribute | `<input ?disabled=${…}>` | Creates a binding that adds or removes the attribute based on truthy/falsey values. |
| JS Property | `<input :value=${…}>` | Creates a binding that sets a JavaScript property on the element. |

> | Event Handler | `<button @click=${…}>...</button>` | Registers an event handler using addEventListener. The listener is automatically removed when the template is unbound. |

Reviewer 6: We should mention that events bound this way are automatically marked with preventDefault unless the event handler returns 'true'.

Author: @scomea Thank you - updated!

| HTML Element Reference | `<button ${ref('myButton')}>...</button>` | Captures a reference to the element and assigns it to the named property on the data source. |
| Slotted Node Capture | `<slot ${slotted('defaultSlotNodes')}></slot>` | Watches the slot for changes and synchronizes those to an array, assigned to the named property on the data source. |

> | Child Node Capture | `<div ${children('divChildren')}></div>` | Watches the element's children or changes and synchronizes those to an array, assigned to the named property on the data source. |

Reviewer 7: data source -> element instance?

Author: @radium-v Thank you - updated!

  template
})
export class MyApp extends FASTElement {
  @observable ready: boolean = false;
  ...
}
```

---

Use the [`repeat`](https://www.fast.design/docs/fast-element/using-directives#the-repeat-directive) directive to render a list of data:

```ts
import { FASTElement, customElement, observable, html, repeat } from '@microsoft/fast-element';

const template = html<FriendList>`
  ...
    ${repeat(x => x.friends, html<string>`
      <li>${x => x}</li>
    `)}
`;

@customElement({
  name: 'friend-list',
  template
})
export class FriendList extends FASTElement {
  @observable friends: Person[] = [];
  ...
}
```

---

Properties available on the context object within a `repeat` block:

| Property | Definition |
| :-- | :-- |
| event | The event object when inside an event handler. |
| parent | The parent view-model when inside a repeat block. |
| parentContext | The parent ExecutionContext when inside a repeat block. This is useful when repeats are nested and the inner-most repeat needs access to the root view-model. |
| index | The index of the current item when inside a repeat block (opt-in). |
| length | The length of the array when inside a repeat block (opt-in). |

> | isEven | True if the index of the current item is even when inside a repeat block (opt-in). |

Reviewer 5: I think we should break apart the `opt-in` properties here. It's been a point of confusion in the past because those properties don't exist sometimes.

If we break this out into two tables, we can identify which properties will *always* exist and which properties will only exist when providing the position configuration. We also show an example of *how* to provide the positioning configuration:

```html
<ul>
  ${repeat(x => x.friends, html<string>`
    <li>${(x, c) => c.index} ${x => x}</li>
  `, { positioning: true })}
</ul>
```

Author: @nicholasrice Great points, thank you - updated!

Author: Would you mind checking my updates on this ^ to see if they make sense?

Reviewer 5: Sorry for the delay, looks great! thank you!

| isOdd | True if the index of the current item is odd when inside a repeat block (opt-in). |
isFirst | True if the current item is first in the array inside a repeat block (opt-in). |
isInMiddle | True if the current item is somewhere in the middle of the array inside a repeat block (opt-in). |
isLast | True if the current item is last in the array inside a repeat block (opt-in). |

---

### [Styles](https://www.fast.design/docs/fast-element/leveraging-css#basic-styles)

`FASTElement` provides a css tagged template helper that allows for the creation of `ElementStyles`.

Adding CSS to a `FASTElement`:

```ts

> import { FASTElement, customElement } from '@microsoft/fast-element';
> import { css } from "@microsoft/fast-element";

Reviewer 7: ```suggestion
import { css, customElement, FASTElement} from '@microsoft/fast-element';
```

Author: @radium-v Thank you - updated!

import { disabledOpacity } from "../design-tokens";

const styles = css`

>   :host([disabled])] {

Reviewer 7: ```suggestion
  :host([disabled]) {
```

Author: @radium-v Thank you - updated!

Use the `cssPartial` tagged template literal to create reusable blocks of partial CSS:

```ts
import { css, cssPartial } from "@microsoft/fast-element";

const partial = cssPartial`color: red;`;
const styles = css`:host{ ${partial} }`;
```
---
### [CSSDirective](https://www.fast.design/docs/fast-element/leveraging-css#cssdirective)

To create a `CSSDirective`, import and extend `CSSDirective` from `@microsoft/fast-element`:

```ts
import { CSSDirective }  from "@microsoft/fast-element"

class RandomWidth extends CSSDirective {}
```

#### [createCSS method](https://www.fast.design/docs/fast-element/leveraging-css#createcss):

`CSSDirective` has a `createCSS()` method that returns a string to be interpolated into an `ElementStyles`:

```ts
class RandomWidth extends CSSDirective {
  createCSS() {
    return "width: var(--random-width);"
  }
}
```

#### [createBehavior method](https://www.fast.design/docs/fast-element/leveraging-css#createbehavior):

The `createBehavior()` method can be used to create a `Behavior` that is bound to the element using the `CSSDirective`:

```ts
class RandomWidth extends CSSDirective {
  private property = "--random-width";
  createCSS() {

>     return `width: var(${this.property});`

Reviewer 7: This example produces an invalid property

Author: @radium-v Thank you - updated!

Author: @radium-v Thank you - updated!  

Would you mind reviewing the updated example?

Also, the code example was taken from the docs.  https://www.fast.design/docs/fast-element/leveraging-css

@EisenbergEffect Would it make sense to update here too in the context of the entire page's content?

![Screen Shot 2021-12-15 at 11 28 38 AM](https://user-images.githubusercontent.com/35910788/146252367-739bcbe6-1359-4f4b-bbf3-5c8f03693853.png)

Reviewer 2: Yes, please.

Author: @EisenbergEffect Thank you - updated!

// create a design token
export const specialColor = DesignToken.create<string>("special-color");
```

[Using design tokens in CSS](https://www.fast.design/docs/design-systems/design-tokens#using-design-tokens-in-css):
```ts
// use the design token to build styles
const styles = css`
  :host {
    background: ${specialColor};
  }
`;

const ancestor = document.querySelector("my-ancestor") as FASTElement;
const descendent = ancestor.querySelector("my-descendent") as FASTElement;
```

[Setting values](https://www.fast.design/docs/design-systems/design-tokens#setting-values):
```ts
// change the value for a given element
specialColor.setValueFor(ancestor, "#FFF");
specialColor.setValueFor(descendent, "#000");
```

[Getting values](https://www.fast.design/docs/design-systems/design-tokens#getting-values):
```ts
// get the value
specialColor.getValueFor(ancestor); // "#FFF"
specialColor.getValueFor(descendent); // "#000"
```

[Deleting values](https://www.fast.design/docs/design-systems/design-tokens#deleting-values):
```ts
// unset the value (inherits from ancestor)
specialColor.deleteValueFor(descendent);
specialColor.getValueFor(descendent); // "#FFF"
```

---

> #### Configuring a FAST Design Token:

Reviewer 5: An important note here is that this design token (and others exported from FAST) are only available from the `@microsoft/fast-components` package because they're part of an implementation of a design system

Author: @nicholasrice Thank you - updated!

**Connect with us**:

- Join our [Discord](https://discord.gg/FcSNfg4) server.
- Report bugs, request features through [Github](https://github.com/Microsoft/fast/issues/new/choose).

**Unsure of what to work on?**

- Here are [good first issues](https://github.com/Microsoft/fast/labels/community:good-first-issue).

---
### [Contributor guide](https://www.fast.design/docs/community/contributor-guide/)
- [Machine setup](https://www.fast.design/docs/community/contributor-guide/#machine-setup)
- [Cloning the repository](https://www.fast.design/docs/community/contributor-guide/#cloning-the-repository)
- [Installing & building](https://www.fast.design/docs/community/contributor-guide/#installing-and-building)
- [Testing](https://www.fast.design/docs/community/contributor-guide/#testing)
- [Submitting a pull request](https://www.fast.design/docs/community/contributor-guide/#submitting-a-pull-request)
- [Merging a pull request](https://www.fast.design/docs/community/contributor-guide/#merging-a-pull-request)

---

### [Contributing to `fast-component`](https://www.fast.design/docs/community/contributor-guide)

```shell
cd packages/web-components/fast-components

> yarn start

Reviewer 7: alternatively, `yarn workspace @microsoft/fast-components start`

Author: @radium-v Thank you

@EisenbergEffect What is your preference?

The docs have `yarn start`

![Screen Shot 2021-12-16 at 1 38 28 PM](https://user-images.githubusercontent.com/35910788/146452925-b3281e91-e2f1-4023-be4d-a3065d5e1665.png)

Reviewer 2: Let's leave it as is for now so I can get this merged and published before heading out for vacation 😄

**Example:**

```ts
import { 
  fastButton, 
  provideFASTDesignSystem 
} from "@microsoft/fast-components";

provideFASTDesignSystem()
  .register(
    fastButton() // custom element registration
  );
```

---
### [@fluentui/web-components](https://docs.microsoft.com/en-us/fluent-ui/web-components/)

**A component library that implements Microsoft's [Fluent Design System](https://www.microsoft.com/design/fluent/#/).**

* Assembles the building blocks of `@microsoft/fast-foundation` to create its component set. 
* Use this library when you want components to look and feel like those found in Microsoft products.

To install the components, use either `npm` or `yarn`:

```shell
npm install --save @fluentui/web-components
``` 

```shell
yarn add @fluentui/web-components
```

To use a Web Component as a custom element in HTML, the custom element must be registered.

**Example:**

```ts
import { 
  fluentButton, 
  provideFluentDesignSystem 
} from "@fluentui/web-components";

provideFluentDesignSystem()
  .register(
    fluentButton() // custom element registration
  );
```

Learn more about [Fluent UI Web](https://github.com/microsoft/fluentui).

> ---

Reviewer: @huynhicode Sorry to nitpick - I don't think we should have anything regarding Fluent in the cheat sheet. Fluent is an implementation of FAST and not related directly outside our involvement. If we can pull any references to Fluent that will likely save us some headache in the long run. This documentation all exists in multiple places on the Fluent side, so I don't think there's anything lost by removing this from here.

Author: @chrisdholt Understandable - I will remove, thank you!

Reviewer 2: While we don't need a full code example here, I think there should be some link over to Fluent as another design system that our team builds. Once we revamp our home page, I think we could remove it, if we add a design systems section to our home page to showcase the systems built on top of FAST. But I think having some sort of link here even for SEO is important. And our community often moves back and forth between FAST and Fluent content.

Reviewer: Is the cheat sheet the right place for that though? We have references to Fluent on the docs as it currently stands and it's duplicative content that seems mostly incidentally related because of our relationship to the project. While our Microsoft team moves back and forth, if the cheat sheet is primarily for the community I think it'll be more confusing than not to tie the two directly together. While I get the value to onboarding someone on our direct team, I think the broader story for the community isn't as clear.

Would comments help here in terms of what things actually are? Or perhaps we can mention that replacing FAST with Fluent in most all cases will work for that repo? I'm just very hesitant to link the two when from a community standpoint they are only related in that one (Fluent) is an implementation of the other.

Author: @EisenbergEffect @chrisdholt Sorry, I didn't see this until too late.  🤦🏻‍♀️ 😞
